### PR TITLE
PP-1188: Add easy valgrind support for PBS

### PR DESCRIPTION
**Pull request by subhasisb**
#### Issue-ID

* **[PP-1188](https://pbspro.atlassian.net/browse/PP-1188)**

#### Problem description
* It is difficult to run valgrind under python and sift through the huge number of leaks and errors listed by valgrind from the Python library as well as PBS Pro code, most of which are false positives, to find the real positives. Thus, a way to get the false positives to zero is very desirable to developers to quickly find problems introduced by their code changes.

#### Cause / Analysis
* No published steps provided along with PBS Pro source code.
* No valgrind suppression file provided to ignore the false python leaks reported by Valgrind.
* Valgrind also detects a bunch of leaks within PBS Pro code which are false positives. There needs to be a way to suppress these in a debug build.

#### Solution description
1. Added instructions to INSTALL describing steps to compile python and PBS Pro for valgrind. Added usage description.

2. Added a standard valgrind leak- suppressions file (to the source) to ignore false leaks seen by valgrind in the Python library.

3. Modify PBS Pro code to remove the false leaks seen by valgrind in the PBS code. Added these in DEBUG mode only so as not to affect performance in non-DEBUG mode.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [ ] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by subhasisb** [comment]
This is the list of PBS Pro source leaks seen by valgrind which are not really leaks.
[val_pbs_code_leaks.txt](https://github.com/PBSPro/pbspro/files/1587582/val_pbs_code_leaks.txt)

**Comment by subhasisb** [comment]
Here is the clean output from valgrind in a simple run of few jobs along with a server periodic hook. 
[val_clean.txt](https://github.com/PBSPro/pbspro/files/1587583/val_clean.txt)

File: src/lib/Libtpp/tpp_transport.c [hunk 1]
```diff
@@ -2304,14 +2326,23 @@ free_phy_conn(phy_conn_t *conn)
...
-
+	free(conn->ctx);
+	free(conn->scratch.data);
+	free(conn->scratch.extra_data);
 	free(conn);
 }
 
+/*
+ * Dummy log function used when terminate is called
+ * We cannot log anything after fork even if tpp_func_func
```
[the remark this answers is not in the document]
Comment (subhasisb) [reply src/lib/Libtpp/tpp_transport.c]: Ah thanks, I will fixed this.

**Comment by subhasisb** [comment]
@arungrover addressed your comments, please check further

**Comment by arungrover** [comment]
I don't think that it is linked, but test_round_robin smoke test is failing on centos

File: src/include/avltree.h [hunk 2]
```diff
@@ -83,6 +83,7 @@ typedef struct {
 #define AVL_DUP_KEYS_OK	1	/* repeated key & rec cause an error message */
 #define AVL_COUNT_DUPS	2	/* complete dups allowed, count repetitions */
 
+extern void *get_avl_tls(void);
```
Comment (ashwathraop) [inline src/include/avltree.h]: Small nit. You might want to reorder the tabs to match with variables beow.

File: valgrind.supp [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+   fun:init_ctypes
+   fun:_PyImport_LoadDynamicModule
+   fun:load_module
+   fun:import_submodule
+   fun:load_next
+   fun:import_module_level
+   fun:PyImport_ImportModuleLevel
+   fun:builtin___import__
+}
+
```
Comment (ashwathraop) [inline valgrind.supp]: In the val_clean.txt you attached at the end of file I see this:
==115610== 17 bytes in 1 blocks are definitely lost in loss record 44 of 1,532
==115610==    at 0x4C29BE3: malloc (vg_replace_malloc.c:299)
==115610==    by 0x67AC949: strdup (strdup.c:42)
==115610==    by 0x472529: main (pbsd_main.c:1984)

Do we need this as well in suppression file?

**Comment by subhasisb** [comment]
Updated val_clean.txt (valgrind output which is now clean)
[val_clean.txt](https://github.com/PBSPro/pbspro/files/1607478/val_clean.txt)

**Comment by bhroam** [comment]
Quick question, did you also suppress the scheduler's python warnings from the formula?  They'll come from using the job_sort_formula or fairshare formula.

**Comment by subhasisb** [comment]
@bhroam no, not yet. My intent so far is to provide an initial valgrind suppresion list that can be augmented over time, if required. For the leaks i could see in the server, i actually did not add them to the valgrind supression file, but removed those "leaks" in DEBUG mode. We do not want to free those memory chunks in production builds, since they simply do not serve any good and actually may even slow down forks/shut-downs etc., so there are many cases where we intentionally do not free memory because the process is either going down or is short lived. However, that becomes a mess for users trying to find real leaks by using valgrind, hence i removed those "intended leaks" DEBUG mode. This way, developers can now build PBS in DEBUG mode and run under valgrind to find genuine leaks. 

Of course, we will find more "non-leaks" in the server/scheduler etc. as we run more tests. We can then decide to either augment the valgrind suppression file or address those leaks in debug mode in the PBS code (or both). 

I believe this will be a continuous work. The point of this PR is to provide a way to make this an easy part of the development process.

**Comment by subhasisb** [comment]
@ashwathraop please check the updated valgrind output attached.

**Comment by subhasisb** [comment]
@arungrover i have approvals now. Squashed and rebased as well.

**Pull request merged**